amm-executor/src/Spectrum/Executor/Backlog/Service.hs

```diff
@@ -1,60 +1,158 @@
   toRevisitQ  <- newIORef Seq.empty -- regular queue; successully submitted orders. Left orders should be re-executed in X minutes. Normally successfully confirmed orders are eliminated from this queue.
+  _           <- liftK $ recover store config pendingPQ
+  pure $ attachLogging logging BacklogService
+    { put = \(PendingOrder order timestamp) -> do
+        put $ BacklogOrder timestamp order
+        modifyIORef pendingPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+    , suspend = \(SuspendedOrder order timestamp) -> do
+        modifyIORef suspendedPQ (PQ.insert $ mkWeightedOrderWithTimestamp order timestamp)
+        isExist $ BacklogOrder timestamp order
+    , checkLater = \(InProgressOrder order timestamp) -> do
+        modifyIORef toRevisitQ (mkWeightedOrderWithTimestamp order timestamp Seq.<|)
+        isExist $ BacklogOrder timestamp order
+    , tryAcquire = do
+        _ <- refreshQueues config store pendingPQ suspendedPQ toRevisitQ
+        maxWeightedOrderM <- getMaxWeightedOrder' pendingPQ suspendedPQ
+        case maxWeightedOrderM of
+          Just (WeightedOrderWithTimestamp oId _ _) -> get oId <&> fmap backlogOrder
+          Nothing -> pure Nothing
+    , drop = \(ExecutedOrder oId) -> do
+        _ <- modifyIORef pendingPQ   (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef suspendedPQ (PQ.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        _ <- modifyIORef toRevisitQ  (Seq.filter (\(WeightedOrderWithTimestamp orderId _ _) -> orderId /= oId))
+        drop oId
+    }
 
-  (_, db) <- Rocks.openBracket storePath
-              Rocks.defaultOptions
-                { Rocks.createIfMissing = createIfMissing
-                }
+refreshQueues
+  :: (MonadIO m)
+  => BacklogServiceConfig
+  -> BacklogStore m
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (PQ.MaxQueue WeightedOrderWithTimestamp)
+  -> IORef (Seq.Seq WeightedOrderWithTimestamp)
+  -> m ()
+refreshQueues BacklogServiceConfig{..} BacklogStore{..} pendingQueueRef suspendedQueueRef toRevisitSeqRef = do
+  currentTime <- getCurrentTime
+  revisited2process <-
+    atomicModifyIORef'
+      toRevisitSeqRef
+      (Seq.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderExecTime))
+  suspended2drop   <-
+    atomicModifyIORef'
+      suspendedQueueRef
+      (PQ.partition (\(WeightedOrderWithTimestamp _ _ oTime) -> diffUTCTime currentTime oTime < orderLifetime))
+  _ <- (\order@(WeightedOrderWithTimestamp oId _ oTime) ->
+          if diffUTCTime currentTime oTime < orderLifetime
+            then modifyIORef' pendingQueueRef (PQ.insert order)
+            else drop oId
+       ) `traverse` revisited2process
+  _ <- (\(WeightedOrderWithTimestamp oId _ _) -> drop oId) `traverse` PQ.toList suspended2drop
```

Comment:
I think expired orders eviction should be done passively during suspended order selection:
```
selectSuspended = do
  ord <- PQ.getMax suspendedQueue
  if isExpired ord
    then drop >> selectSuspended
    else return ord
```

Reply:
It will cause more then one o(n) operation in case when two outdated orders will be on top of queue

Reply:
nope, in case when top order is expired it will take o(n), two orders 2*o(n)

Reply:
Wtf? Why? Where does `n` come from? Each cycle you either pick `max` order from queue or take head from `Seq`, both are constant operations.